### ISSUE TITLE
Incompatible with GoG Version

### ISSUE THREAD

**Issue by mwgabby-li** [gtw123/ShenzhenMod#1]
The GoG version has a different hash and different obfuscated function names, so the patcher won't apply.
Let me know if there's anything I can provide to assist. I was able to run your program from source as well as open the executable with ILSpy. I'm completely new at C# disassembly though.

**Comment by gtw123** [comment]
I can think of a couple of options here:
1. Change the code so that it looks for multiple obfuscated names.
2. Change it to use a different way to identify types. e.g. It could identify a class by the types of its first few fields, or a method by its parameters (and local variables?), or a field by its type and order in the class/struct.

Option 1 would be the most reliable but a pain to maintain because it's probably going to break whenever the game gets an update.

Option 2 sounds a little bit dodgy but as long as the code checks for a unique match I think it should be fine. And it should be much easier to maintain going forward.

I might give option 2 a go later this week and then you could maybe see how it works with the GoG version?

**Comment by gtw123** [comment]
BTW my workflow so far has been something like this:
* Load up the original EXE in ILSpy.
* Use the Reflexil plugin to clean up the names (and more importantly, the strings).
* Search around for something I'm interested in. (Looking for a particular texture name is a good way to start!)
* Use Reflexil to change a number or some code.
* Test it out.
* Add code to ShenzhenMod to do the code changes programatically.

You can also try exporting all the code to a .csproj but unfortunately it either doesn't compile (if you use the obfuscated names) or doesn't export (if you use the cleaned up names).

**Comment by mwgabby-li** [comment]
Since the game is available on Linux and OSX, they're probably using MonoDevelop or something like that.
One of the bonuses of option 2 is that the patcher could probably also be modified to work on those versions, too, not just the Windows version.

**Comment by gtw123** [comment]
As it turns out, the unpatched EXE is actually the same on macOS and Windows, and so the patched EXE already works fine on both OSs! Presumably the macOS version uses Mono to launch the EXE while the Windows version runs it directly.

So if you run the patcher on the Windows version of the game, then copy the patched EXE to the other OS (plus the other files that the patcher installs into Content) then it works fine. I don't have access to a Linux machine but I imagine it would be the same there too. 

Of course it would be much more convenient if the patcher worked directly on all three OSs. Maybe if there is demand for it... :)

**Comment by gtw123** [comment]
I've implemented option 2 now, and the installer/patcher no longer refers to any obfuscated names. There were a few classes/methods that were tricky to "find" but I found a way in the end!

Can you grab the latest source and try running it on the GoG version? You'll need to bypass or update the hash check, but I'm hoping it will then work properly. There are a few places where I hard-coded offsets within methods so they still might not work, but they're not in code that has any Steam references so it might be OK.

Let me know how you go. If it works, could you let me know (a) the hash of the GoG version, and (b) if there is a registry key I could use to locate where it's installed? It might be something like HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\GOG.com, although if you used GoG Galaxy to install it it might be different.

**Comment by mwgabby-li** [comment]
I didn't test it thoroughly, but it seemed to work, and there were no installation warnings.

The hash of the (Windows) GoG version is:
`02-DB-AC-A8-8C-27-4E-A2-8F-A2-7E-D5-92-72-08-9C-F7-5A-DB-B7-5A-88-64-B9-54-05-55-51-7F-6D-56-F6`

The registry path to the SHENZHEN I/O registry metadata for me is:
`HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\GOG.com\Games\1640205738`
I believe I installed it with GoG Galaxy, but I don't remember.

**Comment by opendata26** [comment]
Does this support linux now?

**Comment by mwgabby-li** [comment]
@opendata26 Even if it does, you'll need to run the patcher in Windows on the Linux executable, then copy the executable back to a Linux box.

You could always give it a try if you have access to a Windows box with Visual Studio 2017 (so you can build the patcher and skip the hash check) and the Linux version. Based on what gtw123 said in earlier comments, it will probably work.

**Comment by gtw123** [comment]
@Aradayn OK, I've added proper GoG support now. Could you grab the latest code one more time and check that it now finds and patches the GoG version OK?

To test that the patched version works OK I usually do this:
* Verify there is a "Prototyping Bigger Ideas" item in the puzzle list.
* Make a new solution for this puzzle.
* Add a component in the bottom left and the far top right.
* Draw some traces.
* Go back to the puzzle screen, then open the solution again.
* Make sure everything looks the same as it did before.
* Pick one of the "normal" puzzles and make sure one of your existing solutions still works OK.
* Also, when you edit a solution in the sandbox, the "simulation speed" slider should be twice as wide as normal. Make sure you can drag it all the way to the right without it looking wrong.

If these things all work OK then it's probably all working properly!

**Comment by gtw123** [comment]
@opendata26 Aradayn is right: the patcher itself currently only works on Windows, but I believe it should be able to patch the Linux version too.

If you have a Windows machine (or VM) available, try this:
* Copy the whole Shenzhen I/O folder to the Windows machine.
* Run the mod installer in Windows.
* In the installer, choose the folder you copied to the machine. Specifically, choose the folder containing Shenzhen.exe.
* Click Install.
* If it succeeds, copy the folder back to the Linux machine.  (It patches some additional files besides Shenzhen.exe so it's easiest if you just copy the whole folder.)

That should work for the Steam version. If you have the GoG version you'll need to either (a) build the latest version from source first (in VS2017) or (b) wait until I release it (hopefully tomorrow!).

**Comment by mwgabby-li** [comment]
@gtw123 It almost worked, but the game crashes if you move the simulation slider all the way to the right in the sandbox.
Everything else was fine though. For reference, this was my test puzzle:
[name]
 New design 1
[puzzle] SzSandbox2

[traces] 
............................................
.......................................94...
...................................95556....
..................................96........
................................956.........
...............................96...........
.............................956............
..........................9556..............
........................956.................
.......................96...................
......................96....................
.....................96.....................
.....................A......................
...................956......................
..................96........................
.................96.........................
................96..........................
...............96...........................
.............956............................
............96..............................
...........96...............................
..........96................................
.........96.................................
.......956..................................
.....956....................................
...156......................................
............................................
............................................

[chip] 
[type] UC4X
[x] 1
[y] 1
[code] 
  mov 0 x3
  slp 1

[chip] 
[type] UC4X
[x] 40
[y] 25
[code] 
  mov x0 acc
  slp 1

**Comment by opendata26** [comment]
@Aradayn couldn't I just try and compile with mono.

**Comment by mwgabby-li** [comment]
@opendata26 You're welcome to try to build the patcher with Mono, if that's what you mean, but I have no idea if that will work. Good luck! Maybe if you get it working, you could submit a pull request.

**Comment by gtw123** [comment]
@Aradayn I tried your test program and it crashes the Steam version as well! It was a pretty bad crash too (crashed my display driver).

Here's what I've found so far:
* It seems to be crashing in the DirectX 11 renderer.
* If I remove the method that patches the max game speed, the crash no longer happens.
* It doesn't crash in the regular sandbox, just the bigger one.
* The default speed multiplier is 30 and I was increasing it to 100. I tried a smaller number like 60 but that crashes as well.
* At first I thought it was the trace animation that was crashing, but it crashes even if I click on "hide wires".

So I guess for now I'll probably just disable the max game speed patch.
